File: mobile-app/lib/features/swipe_undo/swipe_undo.dart
/// Pure-dart Helper für Swipe-to-complete + Undo.
/// Kein Flutter-Import — testbar ohne widget-tester.
///
/// Public API:
///  * [inverseMutation] — liefert (type, payload) zur Umkehrung einer Mutation.
///  * [UndoEntry] — ein in der Snackbar offenes Undo-Element.
///  * [UndoStack] — verwaltet zuletzt geswipte Aktionen (single-shot).
library;

/// Berechnet die inverse Mutation, die ein User-Swipe rückgängig macht.
/// TOGGLE_TASK ist idempotent → einfach erneut toggeln.
/// DISMISS_IDEA ↔ REACTIVATE_IDEA. CONVERT_IDEA ↔ REACTIVATE_IDEA.
/// Gibt null zurück, wenn der Typ nicht umkehrbar ist.
({String type, Map<String, dynamic> payload})? inverseMutation(
  String type,
  Map<String, dynamic> payload,
) {
  switch (type) {
    case 'TOGGLE_TASK':
      return (type: 'TOGGLE_TASK', payload: Map<String, dynamic>.from(payload));
    case 'DISMISS_IDEA':
    case 'CONVERT_IDEA':
      return (type: 'REACTIVATE_IDEA', payload: Map<String, dynamic>.from(payload));
    case 'REACTIVATE_IDEA':
      return (type: 'DISMISS_IDEA', payload: Map<String, dynamic>.from(payload));
    default:
      return null;
  }
}

/// Beschreibt einen offenen Undo-Vorgang.
class UndoEntry {
  final String id;
  final String label;
  final String mutationType;
  final Map<String, dynamic> payload;
  final DateTime createdAt;

  UndoEntry({
    required this.id,
    required this.label,
    required this.mutationType,
    required this.payload,
    required this.createdAt,
  });

  /// Inverse Mutation für diesen Eintrag (null falls nicht umkehrbar).
  ({String type, Map<String, dynamic> payload})? get inverse =>
      inverseMutation(mutationType, payload);
}

/// In-Memory Stack mit Verfallszeit. Gerätelokal — wird nicht synchronisiert
/// (analog Theme/UI-State-Regel: Undo ist UI-State).
class UndoStack {
  final Duration ttl;
  final DateTime Function() _now;
  final List<UndoEntry> _entries = [];

  UndoStack({this.ttl = const Duration(seconds: 5), DateTime Function()? now})
      : _now = now ?? DateTime.now;

  void push(UndoEntry e) {
    _entries.add(e);
  }

  /// Aktive Einträge (noch nicht abgelaufen).
  List<UndoEntry> get active {
    final cutoff = _now().subtract(ttl);
    return _entries.where((e) => e.createdAt.isAfter(cutoff)).toList();
  }

  /// Pop by id — gibt Eintrag zurück und entfernt ihn.
  /// null, wenn nicht (mehr) gefunden.
  UndoEntry? consume(String id) {
    final ix = _entries.indexWhere((e) => e.id == id);
    if (ix < 0) return null;
    return _entries.removeAt(ix);
  }

  /// Entferne abgelaufene Einträge. Liefert Anzahl entfernter.
  int purgeExpired() {
    final cutoff = _now().subtract(ttl);
    final before = _entries.length;
    _entries.removeWhere((e) => !e.createdAt.isAfter(cutoff));
    return before - _entries.length;
  }

  int get size => _entries.length;
}

File: desktop-app/swipe_undo.js
/**
 * Pure-JS Helper für Swipe-to-complete + Undo (Desktop-App).
 * Spiegelt mobile-app/lib/features/swipe_undo/swipe_undo.dart.
 * Kein DOM-/React-Import — testbar via node:test.
 *
 * Public API:
 *  - inverseMutation(type, payload): liefert {type, payload} zur Umkehr,
 *    null falls nicht umkehrbar.
 *  - UndoStack: gerätelokaler In-Memory-Stack mit TTL (Default 5s).
 *
 * Hinweis: Undo ist UI-State (analog Theme/UI-Regel) und wird NICHT
 * über den sync_server gespiegelt.
 */

'use strict';

function inverseMutation(type, payload) {
  const p = { ...(payload || {}) };
  switch (type) {
    case 'TOGGLE_TASK':
      return { type: 'TOGGLE_TASK', payload: p };
    case 'DISMISS_IDEA':
    case 'CONVERT_IDEA':
      return { type: 'REACTIVATE_IDEA', payload: p };
    case 'REACTIVATE_IDEA':
      return { type: 'DISMISS_IDEA', payload: p };
    default:
      return null;
  }
}

class UndoStack {
  /**
   * @param {{ttlMs?: number, now?: () => number}} [opts]
   */
  constructor(opts = {}) {
    this.ttlMs = opts.ttlMs ?? 5000;
    this._now = opts.now ?? (() => Date.now());
    this._entries = [];
  }

  /**
   * @param {{id:string,label:string,mutationType:string,payload:object,createdAt?:number}} entry
   */
  push(entry) {
    if (!entry || !entry.id || !entry.mutationType) {
      throw new Error('UndoStack.push: id und mutationType erforderlich');
    }
    this._entries.push({
      id: entry.id,
      label: entry.label || '',
      mutationType: entry.mutationType,
      payload: { ...(entry.payload || {}) },
      createdAt: entry.createdAt ?? this._now(),
    });
  }

  /** Aktive (nicht abgelaufene) Einträge. */
  get active() {
    const cutoff = this._now() - this.ttlMs;
    return this._entries.filter((e) => e.createdAt > cutoff);
  }

  /**
   * Entfernt Eintrag mit id und gibt ihn zurück (null falls nicht (mehr) da).
   * @param {string} id
   */
  consume(id) {
    const ix = this._entries.findIndex((e) => e.id === id);
    if (ix < 0) return null;
    const [removed] = this._entries.splice(ix, 1);
    return removed;
  }

  /** Entfernt abgelaufene Einträge. Gibt Anzahl entfernter zurück. */
  purgeExpired() {
    const cutoff = this._now() - this.ttlMs;
    const before = this._entries.length;
    this._entries = this._entries.filter((e) => e.createdAt > cutoff);
    return before - this._entries.length;
  }

  get size() {
    return this._entries.length;
  }
}

/**
 * Filter: ghost-pending items (__pending===true) dürfen nicht swipe-bar sein.
 * Verhindert, dass offline-queue-eintrag während sync verworfen wird.
 */
function isSwipeable(item) {
  if (!item) return false;
  return item.__pending !== true;
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { inverseMutation, UndoStack, isSwipeable };
}

if (typeof window !== 'undefined') {
  window.swipeUndo = { inverseMutation, UndoStack, isSwipeable };
}
